src/server/routes/oauth.ts

```diff
@@ -0,0 +1,178 @@
+  getAuthorizationStateCookie,
+  OpenIdErrors,
+  ProfileOpenIdClient,
+  ProfileOpenIdClientRedirectUris,
+} from '@/server/lib/okta/openid-connect';
+import { getConfiguration } from '@/server/lib/getConfiguration';
+import { logger } from '@/server/lib/serverSideLogger';
+import { trackMetric } from '@/server/lib/trackMetric';
+import { handleAsyncErrors } from '@/server/lib/expressWrappers';
+import { exchangeAccessTokenForCookies } from '@/server/lib/idapi/auth';
+import { setIDAPICookies } from '@/server/lib/setIDAPICookies';
+import { SignInErrors } from '@/shared/model/Errors';
+import {
+  readEncryptedStateCookie,
+  setEncryptedStateCookie,
+} from '@/server/lib/encryptedStateCookie';
+import { addQueryParamsToPath } from '@/shared/lib/queryParams';
+
+interface OAuthError {
+  error: string;
+  error_description: string;
+}
+
+const { defaultReturnUri } = getConfiguration();
+
+/**
+ * Type guard to check that a given error is an OAuth error.
+ * By checking for the `error` and `error_description` properties
+ * @param {unknown} obj
+ * @return {boolean}
+ */
+const isOAuthError = (
+  maybeOAuthError: unknown,
+): maybeOAuthError is OAuthError => {
+  const { error, error_description } = maybeOAuthError as OAuthError;
+  return error !== undefined && error_description !== undefined;
+};
+
+/**
+ * Helper method to redirect back to the sign in page with
+ * a generic error message if we don't want to expose the error
+ * back to the client. Be sure to log the error though!
+ */
+const redirectForGenericError = (
+  req: Request,
+  res: ResponseWithRequestState,
+) => {
+  // we set the sign in redirect to true so that we can render the sign in page
+  // without getting into a redirect loop
+  const encryptedState = readEncryptedStateCookie(req);
+  setEncryptedStateCookie(res, {
+    ...encryptedState,
+    signInRedirect: true,
+  });
+
+  return res.redirect(
+    addQueryParamsToPath('/signin', res.locals.queryParams, {
+      error_description: SignInErrors.GENERIC,
+    }),
+  );
+};
+
+router.get(
+  '/oauth/authorization-code/callback',
+  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+    try {
+      // params returned from the /authorize endpoint
+      // for auth code flow they will be "code" and "state"
+      // "code" is the authorization code to exchange for access token
+      // "state" will be the "nonce" value set in the oidc_auth_state cookie
+      // if there were any errors, then an "error", and "error_description" params
+      // will be returned instead
+      const callbackParams = ProfileOpenIdClient.callbackParams(req);
+
+      // get the oidc_auth_state cookie which contains the "nonce"
+      // and "returnUrl" so we can get the user back to the page they
+      // initially landed on sign in from
+      const authState = getAuthorizationStateCookie(req);
+
+      // check if the state cookie exists
+      if (!authState) {
+        // no state! is this an attack?
```

Comment:
the `authState` would have to be set by gateway before performing the redirect to `/authorize`, which happens here: https://github.com/guardian/gateway/pull/1410/files#diff-88981251923a306ae51df84b8db7eae8eab850112e93512e07c21fc7c151dccbR54

If this doesn't exist, that would mean that either a) the state isn't being set correctly, or b) someone is trying to attack the oauth flow for example with an invalid state cookie, or without the state cookie

basically the state parameter is used to prevent CSRF attacks, if the state parameter is missing, or invalid, or different it could mean that someone is trying those kinds of attacks.

i'll update the comments

Reply:
thanks, the extra info is really helpful :)